sklearn/utils/estimator_checks.py

```diff
@@ -345,28 +384,33 @@ def _generate_class_checks(Estimator):
     yield from _generate_instance_checks(name, estimator)
 
 
-def _mark_xfail_checks(estimator, check, pytest):
+def _generate_checks(Estimator):
+    if isinstance(Estimator, type):
+        # got a class
+        checks_generator = _generate_class_checks(Estimator)
+    else:
+        # got an instance
+        estimator = Estimator
+        name = type(estimator).__name__
+        checks_generator = _generate_instance_checks(name, estimator)
+
+    return checks_generator
+
+
+def _generate_marked_checks(estimator, pytest):
     """Mark (estimator, check) pairs with xfail according to the
     _xfail_checks_ tag"""
-    if isinstance(estimator, type):
-        # try to construct estimator instance, if it is unable to then
-        # return the estimator class, ignoring the tag
-        try:
-            estimator = _construct_instance(estimator)
-        except Exception:
-            return estimator, check
-
-    xfail_checks = estimator._get_tags()['_xfail_checks'] or {}
-    check_name = _set_check_estimator_ids(check)
+    xfail_checks_tag = _get_xfail_checks(estimator)
+    checks_generator = _generate_checks(estimator)
 
-    if check_name not in xfail_checks:
-        # check isn't part of the xfail_checks tags, just return it
-        return estimator, check
-    else:
-        # check is in the tag, mark it as xfail for pytest
-        reason = xfail_checks[check_name]
-        return pytest.param(estimator, check,
-                            marks=pytest.mark.xfail(reason=reason))
+    for estimator, check in checks_generator:
+        check_name = _set_check_estimator_ids(check)
+        reason = xfail_checks_tag.get(check_name, "")
+        if reason:
```

Comment:
`if reason:...` doesn't immediately read as "if the check is part of the xfails tag"

I personally prefer the previous logic of explicitly checking whether check_name is present in `xfail_checks`. That's one of the things that lead me to this refactoring PR the other day